### ISSUE TITLE
Incorrect parameter name in policy_set_definition.policy_definition_reference

### ISSUE THREAD

**Issue by jonpmay** [terraform-providers/terraform-provider-azurerm#10155]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
→ tf -version
Terraform v0.14.4
+ provider registry.terraform.io/hashicorp/azurerm v2.42.0
```

### Affected Resource(s)

* `azurerm_policy_set_definition`

### Terraform Configuration Files

```hcl
resource "azurerm_policy_set_definition" "abc" {
  name                  = "foo"
  display_name          = "bar"
  policy_type           = "Custom"
  management_group_name = azurerm_management_group.xyz.name

  policy_definition_group {
    name         = "foo"
    display_name = "bar"
  }

  policy_definition_reference {
    policy_definition_id = "/providers/Microsoft.Authorization/policyDefinitions/{GUID}"
    group_names = ["foo"]
  }
}
```

### Expected Behaviour

According to the [azurerm_policy_set_definition documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/policy_set_definition#policy_definition_reference), the `policy_definition_reference` block defines `policy_group_names` as an available parameter.

This parameter name can be confirmed in the [azurerm_policy_set_definition](https://github.com/terraform-providers/terraform-provider-azurerm/blob/678402fa93be8e375e9ab676c009349dfe7a9bd5/azurerm/internal/services/policy/policy_set_definition_resource.go#L158-L163) object code.

### Actual Behavior

According to the [azurerm_policy_set_definition documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/policy_set_definition#policy_definition_reference), the `policy_definition_reference` block defines `group_names` as an available parameter.

```
terraform validate

Error: Unsupported argument

  on main.tf line 48, in resource "azurerm_policy_set_definition" "abc":
  48:     group_names = ["foo"]

An argument named "group_names" is not expected here.
```

### Steps to Reproduce

1. Create `azurerm_policy_set_definition` resource with a `policy_definition_reference` block and include the `group_names` parameter
2. `terraform init`
3. `terraform plan`

**Comment by fgarcia-cnb** [comment]
im seeing this as well

**Comment by HarleyB123** [comment]
Looks like an error with the documentation rather than a Terraform limitation. The parameter should be **policy_group_names** , see [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/policy/policy_set_definition_resource.go#L157-L164).

Changing __group_names__ to __policy_group_names__ works for me

**Comment by fgarcia-cnb** [comment]
I made the change to policy_group_names and the apply completes successfully now, however the apply isn't actually making changes to an existing policy set. Policy groups seem to only be created properly when creating the policy set, but not on changes post creation (i.e. renaming a policy group, adding policy groups, assigning a policy to a different group)

for example i have the following:
```hcl
policy_definition_groups = ["INFRA-TEST-Initiative3-test-group1", "INFRA-TEST-Initiative3-test-group2"]
```

the policy set creates it successfully and assigns policies to their assigned group. if i then set the groups to:
```hcl
policy_definition_groups = ["INFRA-TEST-Initiative3-test-group1", "INFRA-TEST-Initiative3-test-group2", "INFRA-TEST-Initiative3-test-group3"]
```

a plan and apply makes no changes, and keeps showing the following:
```hcl
      - policy_definition_group {
          - name = "INFRA-TEST-Initiative3-test-group1" -> null
        }
      - policy_definition_group {
          - name = "INFRA-TEST-Initiative3-test-group2" -> null
        }
      + policy_definition_group {
          + name = "INFRA-TEST-Initiative3-test-group3"
        }
      + policy_definition_group {
          + name = "INFRA-TEST-Initiative3-test-group1"
        }
      + policy_definition_group {
          + name = "INFRA-TEST-Initiative3-test-group2"
        }
```

its strange that its redoing all the groups instead of just appending the new one. also, it cleared out the existing policy group assignments. the assignment below existed on policy set creation, but is no longer there (and it keeps trying to apply unsuccessfully):

```hcl
~ policy_definition_reference {
          + parameter_values     = jsonencode({})
            parameters           = {}
            policy_definition_id = "/providers/Microsoft.Authorization/policyDefinitions/06a78e20-9358-41c9-923c-fb736d382a4d"
          ~ policy_group_names   = [
              + "INFRA-TEST-Initiative-test-group4-2",
            ]
            reference_id         = "2"
        }
```

**Comment by HarleyB123** [comment]
Hey @fgarcia-cnb 

Sorry didn't get back to you earlier on this. Have done some digging and this is why I think that issue occurs:

On the create function [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/policy/policy_set_definition_resource.go#L316), we can see that it calls ```expandAzureRMPolicySetDefinitionPolicyDefinitions```.  Sure enough, this function returns the GroupNames on create [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/0134b52302748e6e4a04ac1b011fa6b46d8d9c4a/azurerm/internal/services/policy/policy_set_definition_resource.go#L659). As you've said, creation works fine in this case 😄 

For updating the resource, the provider calls a slightly different function [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/0134b52302748e6e4a04ac1b011fa6b46d8d9c4a/azurerm/internal/services/policy/policy_set_definition_resource.go#L444), called ```expandAzureRMPolicySetDefinitionPolicyDefinitionsUpdate```. From the result [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/0134b52302748e6e4a04ac1b011fa6b46d8d9c4a/azurerm/internal/services/policy/policy_set_definition_resource.go#L622-L626), we can see that GroupNames isn't returned.  I think just editing the return value on the update function should fix this issue:

```go
result = append(result, policy.DefinitionReference{
        PolicyDefinitionID:          utils.String(d.Get(fmt.Sprintf("policy_definition_reference.%d.policy_definition_id", i)).(string)),
        Parameters:                  parameters,
	PolicyDefinitionReferenceID: utils.String(d.Get(fmt.Sprintf("policy_definition_reference.%d.reference_id", i)).(string)),
	GroupNames:                  utils.ExpandStringSlice(v["policy_group_names"].(*schema.Set).List()),
        })
}
```

Anyone know whether this is correct before I make a PR? 👍🏻

**Comment by fgarcia-cnb** [comment]
much appreciated @HarleyB123. I was wondering, if you have time, maybe you can look at these other issues with policy_set_definition? #6072 

these bugs make the resource basically unusable. I've made some recent posts with useful info, but no one else has looked at the issue in about a year

**Comment by AnthonyRocchio** [comment]
Hello, we're seeing this as well regarding loss of group assignments to policies in a policy set.  Initial creation of the policy set assigns groups to policies appropriately.  However, upon any subsequent update to the policies in the set (ex - removal or addition of policies to/from policy set) all groups assigned to policies are blanked in Azure (though as fgarcia-cnb states, terrform does keep trying to apply the groups).  Using Azurerm 2.61.0.  Thanks!

**Comment by HarleyB123** [comment]
Hey - I've given it a go at solving this, put a PR in place to fix the updates - It still wants to recreate all the policy groups like @fgarcia-cnb 's example shows, but in my PR these changes are actually applied rather than wiped (feel free to clone down my PR of the provider and give it a try)!

**Comment by katbyte** [comment]
closed by #12279
